core/src/main/scala/org/apache/spark/scheduler/Task.scala

```diff
@@ -155,7 +155,13 @@ private[spark] abstract class Task[T](
    */
   def collectAccumulatorUpdates(taskFailed: Boolean = false): Seq[AccumulatorV2[_, _]] = {
     if (context != null) {
-      context.taskMetrics.accumulators().filter { a => !taskFailed || a.countFailedValues }
+      context.taskMetrics.internalAccums.filter { a =>
+        // RESULT_SIZE accumulator is always zero at executor, we need to send it back as its
+        // value will be updated at driver side.
+        !a.isZero || a.name == Some(InternalAccumulator.RESULT_SIZE)
+      // zero value external accumulators may still be useful, e.g. SQLMetrics, we should not filter
```

Comment:
There are 2 concepts:

1. internal accumulators: like GCtime, resultSize, which are internal to DAGScheduler.
2. `countFailedValues` accumulator: `countFailedValues` is an internal flag that can only be set by us. All internal accumulators are `countFailedValues` accumulators, and SQLMetrics are also `countFailedValues` accumulators.